oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java

```diff
@@ -362,5 +382,17 @@ private String buildAuthorizationRequestUri() {
 					.build()
 					.toUriString();
 		}
+
+		private String getCodeChallenge(String codeVerifier) {
```

Comment:
Just as the OAuth2AuthorizationRequest has no logic in generating the state or code verifier, I'd prefer to avoid logic on generating a code challenge. More concretely, I think we should externalize how the code challenge is created rather than embedding the logic in here. This gives us more flexibility for supporting other types of challenges.

Reply:
Agreed. See my comment on putting this logic in `DefaultOAuth2AuthorizationRequestResolver` instead

Reply:
I'm open to change on this but I feel strongly that the Code Challenge shouldn't be calculated outside of the OAuth2AuthorizationRequest. The Code Challenge is a function of the Code Verifier such that if you know the Code Verifier and the Code Challenge Method (i.e. "f(x)"--where "f()" is always SHA-256 unless the client can't calculate it and "x" is the Code Verifier) then you can determine the Code Challenge. It feels wrong (to me) to store both the Code Verifier AND the Code Challenge on the OAuth2AuthorizationRequest. By setting the Code Challenge externally (i.e. in the DefaultOAuth2AuthorizationRequestResolver), then the users of the OAuth2AuthorizationRequest.Builder could set incompatible Code Verifier and Code Challenge values. That would seems like a bad builder api to me. I'm ok breaking out the actual transformation/function to another class with the ability to set a future new transformation but even that seems unnecessary at this point with the RFC being what it is (i.e. S256 mandatory unless client can't->PLAIN). Again, I'm open to change my mind but can you see where I'm coming from? I understand wanting to keep logic out of what's essentially an stateful entity but I don't see it so much as "logic"--rather a straightforward mathematical function of the Code Verifier. Please convince me otherwise. Thanks!

Reply:
The `OAuth2AuthorizationRequest` is a domain class, pretty much a POJO but OAuth2 specific.  In my view, it stores attributes of an Authorization Request, e.g. `client-id`, `redirect-uri`, `response_type`, etc. As far as PKCE support goes, it also stores attributes such as `code_challenge` and `code_challenge_method`. These attributes serve as standard and additional (`OAuth2AuthorizationRequest.additionalParameters`) request parameters to the Authorization Endpoint. 

The `DefaultOAuth2AuthorizationRequestResolver` is responsible for creating an `OAuth2AuthorizationRequest` and setting it's attributes/parameters. It makes sense to me that the `code_challenge` and `code_challenge_method` would also be set in `OAuth2AuthorizationRequest.additionalParameters` by `DefaultOAuth2AuthorizationRequestResolver`. However, before it can compute the `code_challenge`, it needs to create the `code_verifier` and associate it to `OAuth2AuthorizationRequest` so it can be sent later to the Token Endpoint. It doesn't make sense to store the `code_verifier` in `additionalParameters()` because it's not a parameter to the Authorization Endpoint. So I'm going to propose that the `code_verifier` is stored in `OAuth2AuthorizationRequest.attributes` (see related #5940). Th `attributes` member serves the purpose of storing contextual information of an Authorization Request, which the `code_verifier` is exactly that as it lives during the duration of the Authorization Request and subsequent callback of the Authorization Response.

What are your thoughts @sdoxsee @rwinch ?

Reply:
Hey @jgrandja. I agree with your thinking on this. The named parameters on the OAuth2AuthorizationRequest should really be about the authorization endpoint call and not the token endpoint which would make something like `attributes` good for the `code_verifier`--to be used for the token endpoint call. Once the PKCE parameters are moved into additionalParameters and attributes to keep the PKCE specifics out entirely, I can buy the proposition of doing the `code_challenge` logic in the DefaultOAuth2AuthorizationRequestResolver. I like it because it's extensible but keeps things simple. I don't like subclassing if I can avoid it.

Several more questions:

1. Can I use PkceParameterNames as your solution means that OAuth2AuthorizationRequest is agnostic of PKCE and I don't like dropping them into OAuth2ParameterNames?
2. Shall I mix issues and do #5940 as well?
3. @rwinch are you ok with this before I code it up?

Reply:
Thanks @jgrandja. If you're able to wrap #5940 up quickly then that might be best for efficiency--I'll rebase when you let me know. I'm ok with OAuth2ParameterNames or PkceParameterNames. I suppose the same could be argued for the existing OidcParameterNames though as well? I'll wait for @rwinch before coding anyway.

Reply:
Ok I'll take care of #5940 and let you know when it's in. Good point re: `OidcParameterNames` - let's see what @rwinch has to say.

Reply:
@sdoxsee #5940 has been resolved via 594a169